File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {

  static int[][] direction = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };

  public int solution(int[][] maps) {

    int N = maps.length;
    int M = maps[0].length;

    int[][] visited = new int[N][M];

    bfs(maps, visited, N, M);

    int answer = visited[N - 1][M - 1];

    return (answer > 0) ? answer : -1;
  }

  static void bfs(int[][] maps, int[][] visited, int N, int M) {
    int X = 0;
    int Y = 0;
    visited[Y][X] = 1;
    Queue<int[]> queue = new LinkedList<>();
    queue.add(new int[] { Y, X });

    while (!queue.isEmpty()) {
      int[] now = queue.poll();
      int R = now[0];
      int C = now[1];

      for (int i = 0; i < 4; i++) {
        int DR = R + direction[i][0];
        int DC = C + direction[i][1];

        if (DR < 0 || DR >= N || DC < 0 || DC >= M)
          continue;

        if (visited[DR][DC] == 0 && maps[DR][DC] == 1) {
          queue.add(new int[] { DR, DC });
          visited[DR][DC] = visited[R][C] + 1;
        }
      }
    }
  }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    w = len(maps[0])
    h = len(maps)
    visited = [[0] * (w + 1) for _ in range(h + 1)]
    
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    queue = deque()
    queue.append((0,0))
    visited[0][0] = 1
    while queue :
        x,y = queue.popleft()
        
        if x == h - 1 and y == w - 1 :
            answer = visited[x][y]
            break
        
        for i in range(4) :
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0<=nx<h and 0<=ny<w :
                if not visited[nx][ny] and maps[nx][ny] == 1 :
                    visited[nx][ny] = visited[x][y] + 1
                    queue.append((nx,ny))
                    
    return answer if answer >0 else -1